eth/fetcher/tx_fetcher.go

```diff
@@ -186,134 +251,257 @@ func (f *TxFetcher) Stop() {
-					continue
+			// If any peers became active and are idle, request transactions from them
+			if len(actives) > 0 {
+				f.scheduleFetches(timeoutTimer, actives)
+			}
+
+		case <-timeoutTimer.C:
+			// Clean up any expired retrievals and avoid re-requesting them from the
+			// same peer (either overloaded or malicious, useless in both cases). We
+			// could also penalize (Drop), but there's nothing to gain, and if could
+			// possibly further increase the load on it.
+			for peer, req := range f.requests {
+				if time.Since(req.time)+gatherSlack > txFetchTimeout {
+					// Reschedule all the not-yet-delivered fetches to alternate peers
+					delete(f.requests, peer)
+					for _, hash := range req.hashes {
+						if _, ok := f.announced[hash]; ok {
+							panic("announced tracker already contains alternate item")
+						}
+						if f.alternates[hash] != nil { // nil if tx was broadcast during fetch
+							f.announced[hash] = f.alternates[hash]
+						}
+						delete(f.alternates, hash)
+						delete(f.fetching, hash)
+					}
+					txFetchTimeoutMeter.Mark(int64(len(req.hashes)))
+				}
+			}
+			// Schedule a new transaction retrieval
+			f.scheduleFetches(timeoutTimer, nil)
+
+		case delivery := <-f.cleanup:
+			// Independent if the delivery was direct or broadcast, remove all
+			// traces of the hash from internal trackers
+			for _, hash := range delivery.hashes {
+				delete(f.waitlist, hash)
+				delete(f.waittime, hash)
+				delete(f.announced, hash)
+				for _, txset := range f.announces {
+					delete(txset, hash)
 				}
-				txFetchDurationTimer.UpdateSince(anno.time)
-				txFetchSuccessMeter.Mark(1)
 				delete(f.fetching, hash)
+				delete(f.alternates, hash)
+			}
+			// In case of a direct delivery, also reschedule anything missing
+			// from the original query
+			if delivery.direct {
+				// Make sure something was pending, nuke it
+				req := f.requests[delivery.origin]
+				if req == nil {
+					log.Warn("Unexpected transaction delivery", "peer", delivery.origin)
+					break
+				}
+				delete(f.requests, delivery.origin)
+
+				txFetchDurationTimer.UpdateSince(req.time)
+				txFetchSuccessMeter.Mark(int64(len(delivery.hashes)))
+
+				// Anything not delivered should be re-scheduled (with or without
+				// this peer, depending on the response cutoff)
+				delivered := make(map[common.Hash]struct{})
+				for _, hash := range delivery.hashes {
+					delivered[hash] = struct{}{}
+				}
+				var cutoff int
+				for i, hash := range req.hashes {
+					if _, ok := delivered[hash]; ok {
+						cutoff = i
+					}
+				}
+				// Reschedule missing hashes from alternates, not-fulfilled from alt+self
+				for i, hash := range req.hashes {
+					if _, ok := delivered[hash]; !ok {
+						if i > cutoff {
+							f.alternates[hash][delivery.origin] = struct{}{}
+						}
+					}
+					if _, ok := f.announced[hash]; ok {
+						panic("announced tracker already contains alternate item")
```

Comment:
I find this odd. Let's say `req.hashes` is `[h1,h2]`. I deliver `[h3, h4]`. You will purge `f.announced[h3]` and `f.announced[h4]` on line 363, but afaict this `panic` will hit, because now you're iterating over `req` and not `delivery`

Reply:
The panic here really is a sanity check that every hash exists only either in the waitlist (waiting to time out to be fetched); either in the announced list (waiting to be shceudled for fetching); or the fetching sets.

Re your example, if I request A and B, and you deliver C and D, I'll nuke out all notions of C and D, but since A and B is missing, those muse only exist in f.alternates (the fethcing sets), which I move back into the queues.